core/modules/replica/DatabaseMySQLRow.cc

```diff
@@ -0,0 +1,174 @@
+    }
+    return  _index2cell.size();
+}
+
+bool Row::isNull(size_t              columnIdx) const { return not getDataCell(columnIdx) .first; }
+bool Row::isNull(std::string const& columnName) const { return not getDataCell(columnName).first; }
+
+bool Row::get(size_t             columnIdx,  std::string& value) const { return ::getAsString(*this, columnIdx,  value); }
+bool Row::get(std::string const& columnName, std::string& value) const { return ::getAsString(*this, columnName, value); }
+
+bool Row::get(size_t columnIdx, uint64_t& value) const { return ::getAsNumber(*this, columnIdx, value); }
+bool Row::get(size_t columnIdx, uint32_t& value) const { return ::getAsNumber(*this, columnIdx, value); }
+bool Row::get(size_t columnIdx, uint16_t& value) const { return ::getAsNumber(*this, columnIdx, value); }
+bool Row::get(size_t columnIdx, uint8_t&  value) const { return ::getAsNumber(*this, columnIdx, value); }
+
+bool Row::get(std::string const& columnName, uint64_t& value) const { return ::getAsNumber(*this, columnName, value); }
+bool Row::get(std::string const& columnName, uint32_t& value) const { return ::getAsNumber(*this, columnName, value); }
+bool Row::get(std::string const& columnName, uint16_t& value) const { return ::getAsNumber(*this, columnName, value); }
+bool Row::get(std::string const& columnName, uint8_t&  value) const { return ::getAsNumber(*this, columnName, value); }
+
+bool Row::get(size_t columnIdx, int64_t& value) const { return ::getAsNumber(*this, columnIdx, value); }
+bool Row::get(size_t columnIdx, int32_t& value) const { return ::getAsNumber(*this, columnIdx, value); }
+bool Row::get(size_t columnIdx, int16_t& value) const { return ::getAsNumber(*this, columnIdx, value); }
+bool Row::get(size_t columnIdx, int8_t&  value) const { return ::getAsNumber(*this, columnIdx, value); }
+
+bool Row::get(std::string const& columnName, int64_t& value) const { return ::getAsNumber(*this, columnName, value); }
+bool Row::get(std::string const& columnName, int32_t& value) const { return ::getAsNumber(*this, columnName, value); }
+bool Row::get(std::string const& columnName, int16_t& value) const { return ::getAsNumber(*this, columnName, value); }
+bool Row::get(std::string const& columnName, int8_t&  value) const { return ::getAsNumber(*this, columnName, value); }
+
+bool Row::get(size_t columnIdx, float&  value) const { return ::getAsNumber(*this, columnIdx, value); }
+bool Row::get(size_t columnIdx, double& value) const { return ::getAsNumber(*this, columnIdx, value); }
+
+bool Row::get(std::string const& columnName, float&  value) const { return ::getAsNumber(*this, columnName, value); }
+bool Row::get(std::string const& columnName, double& value) const { return ::getAsNumber(*this, columnName, value); }
+
+bool Row::get(size_t columnIdx, bool& value) const {
+    uint8_t number;
+    if (::getAsNumber(*this, columnIdx, number)) {
+        value = number != '0';
```

Comment:
Not sure what this does. You are comparing number with character '0', how does it work?

Does this handle BIT(1) database type too (I'm not sure how it is returned back to client)?

Reply:
Something is confusing here. Should not `getAsNumber()` make a number `0` of that `"0"` string? Or does `getAsNumber()` work differently for `uint8_t` and assumes it's a character type. If latter then you should probably make it more explicit and use `getAsString()` instead to avoid any unnecessary conversions.

In mysql there is no actual boolean type, I think their `BOOLEAN` is just an alias for `TINYINT(1)`. Some people use `BIT(1)` to represent booleans and I think for that type mysql returns `"\0"` or `"\1"` for values of that type (bit strings). I'm not sure how this code would handle that.

Reply:
I don't remember an exact reason why I chose to use `getAsNumber`. I gather this was done because of a specific behavior of `boost::lexical_cast<T>` for type `uint8_t` (which I had to use for `BOOLEAN/TYNUINT(1)`) exactly like you're speculating above. I think `getAsString` should also work here. Though, it would be a bit less efficient compared with the current method. I had no intent to support all paste of MySQL types (`BIT(1)`) in this implementation because I had no intent to build a general-purpose OO API for MySQL. Only those types which are used by the Replication system are guaranteed to be supported. If you think we should come up with such a general purpose library then we should probably work together on designing/implementing the one. I'm all for it!